src/utils/common.py

```diff
@@ -128,14 +128,49 @@ def compute_ctc_loss(
     targets = torch.LongTensor([char2int_dict.get(c, char2int_dict["UNK"]) for c in chars])
 
     get_len = np.vectorize(len)
-    target_lengths = pad_sequence([torch.from_numpy(get_len(arr)) for arr in texts], batch_first=True, padding_value=0)
-
+    #target_lengths = pad_sequence([torch.from_numpy(get_len(arr)) for arr in texts], batch_first=True, padding_value=0)
+    target_lengths = []
+    for arr in texts:
+        if not len(arr):
+            continue
+        target_lengths.append(torch.from_numpy(get_len(arr)))
+    target_lengths = torch.cat(target_lengths, dim=0)
     ctc_loss = criterion(log_probs, targets, input_lengths, target_lengths)
     ctc_loss /= len(texts)
 
     return ctc_loss
 
 
+def masking(num_patches, ratio=0.25):
+    masked_patches_idx = set()
+    neighbours_idx = set()
+    failures = 0
+    mask = np.zeros(shape=num_patches, dtype=np.int)
+
+    max_masked_patches = math.floor(num_patches * ratio)
+    if max_masked_patches == 0:
+        return mask
+
+    while len(masked_patches_idx) < max_masked_patches and failures < 100:
+        mask_len = random.randint(1, min(6, max_masked_patches))
+        left_patch = random.randint(0, num_patches - mask_len)
+        right_patch = left_patch + mask_len
+
+        cur_masked = set(range(left_patch, right_patch))
+        cur_neighbours = set(range(left_patch - mask_len, left_patch)).union(range(right_patch, right_patch + mask_len))
+
+        if not cur_masked.intersection(neighbours_idx) and not cur_neighbours.intersection(masked_patches_idx):
+            masked_patches_idx.update(cur_masked)
+            neighbours_idx.update(cur_neighbours)
+            failures = 0
+        else:
+            failures += 1
+
+    mask[list(masked_patches_idx)] = 1
+
+    return mask
```

Comment:
That is very bad and slow way to create mask :(

First of all, you need to think in matricies. Let this function take batch size and seq len as arguments.
Secondly, max span is an important parameter, it must also be passed as an argument.

```python
def create_noise_mask(batch_size, seq_len, noise_density=0.2, span_lens=6):
    # 1. Create random mask
    rand_nums = torch.rand(batch_size, seq_len)
    mask = rand_nums.le(noise_density)

    # 2. Calculate position of each mask in span
    cum_sum = torch.cumsum(mask, dim=-1)
    cum_sum_not_mask = torch.masked_fill(cum_sum, mask, 0)
    mask_position = cum_sum - torch.cummax(cum_sum_not_mask, dim=-1).values

    # 3. Avoid too long spans
    mask[mask_position > span_lens] = False
    return mask
```

Please, double check my code. It is good to write here stress tests, e.g. generate 1M mask and natively check all conditions.